bootstrap/0036_sql_tools.sql

```diff
@@ -0,0 +1,154 @@
+    let insert_query_expr text := (select array_to_string(:insert_query_cols, ', '));
+
+    let merge_tmpl varchar := $$MERGE INTO internal.{table} dest USING (
+SELECT {using_expr} FROM (SELECT PARSE_JSON(?) AS obj)) src
+ON dest.{key} = src.{key}
+WHEN MATCHED THEN
+    UPDATE SET {update_expr}
+WHEN NOT MATCHED THEN
+    INSERT ({insert_target_expr}) values ({insert_query_expr})
+$$;
+    let merge_stmt varchar := (select tools.templatejs(:merge_tmpl,
+        {'table': :tbl, 'using_expr': :using_expr, 'key': :key, 'update_expr': :update_expr,
+            'insert_target_expr': :insert_target_expr, 'insert_query_expr': :insert_query_expr}));
+    execute immediate :merge_stmt using (obj);
+end;
+
+CREATE OR REPLACE FUNCTION VALIDATION.IS_NULL(name text)
+    RETURNS text
+    COMMENT='Validates that a field in an object is null'
+AS
+$$
+    'TO_CHAR(f:' || name || ') is null'
+$$;
+
+CREATE OR REPLACE FUNCTION VALIDATION.NOT_NULL(name text)
+    RETURNS text
+    COMMENT='Validates that a field in an object is not null'
+AS
+$$
+    'TO_CHAR(f:' || name || ') is not null'
+$$;
+
+CREATE OR REPLACE FUNCTION VALIDATION.LABEL_ABSENT()
+    RETURNS text
+AS
+$$
+    '(select count(*) = 0 from INTERNAL.LABELS where group_name is null and name = f:name)'
+$$;
+
+CREATE OR REPLACE FUNCTION VALIDATION.GROUPED_LABEL_ABSENT()
+    RETURNS TEXT
+AS
+$$
+    '(select count(*) = 0 from INTERNAL.LABELS where group_name = f:group_name and name = f:name)'
+$$;
+
+CREATE OR REPLACE FUNCTION VALIDATION.GROUP_NAME_UNIQUE()
+    RETURNS TEXT
+AS
+$$
+    '(select count(*) = 0 from internal.labels where name = f:group_name and group_name is null)'
+$$;
+
+CREATE OR REPLACE PROCEDURE VALIDATION.VALID_LABEL_CONDITION(input varchar)
+    RETURNS TABLE(result BOOLEAN)
+    LANGUAGE SQL
+AS
+BEGIN
+    let c varchar := (select parse_json(:input):condition);
+    let stmt varchar := (select 'select ' || :c || ' from reporting.enriched_query_history limit 1');
+    execute immediate :stmt;
+    -- if the condition is invalid, we let the exception propagate.
+    let rs resultset := (select true);
+    return table(rs);
+END;
+
+CREATE OR REPLACE PROCEDURE VALIDATION.IS_VALID_LABEL_NAME(input varchar)
+    RETURNS TABLE(result BOOLEAN)
+    LANGUAGE SQL
+    comment='Validates that the label [group] name does not exist as a column in the reporting.enriched_query_history table'
+AS
+BEGIN
+    let o object := (select parse_json(:input));
+    let n varchar := (select COALESCE(TO_CHAR(:o['group_name']), :o['name']) as name);
+    let stmt varchar := (select 'select ' || :n || ' from reporting.enriched_query_history where false');
+    execute immediate stmt;
+    let rs resultset := (select false);
+    return table(rs);
+EXCEPTION
+    WHEN STATEMENT_ERROR THEN
+        let rs resultset := (select true);
+        return table(rs);
+    WHEN OTHER then
+        let rs resultset := (select false);
+        return table(rs);
```

Comment:
I had to change this proc and the previous one to returning a table so https://github.com/sundeck-io/OpsCenter/pull/591/files#diff-4d8db5ec382ac980083e32c616facb93597373fbc86135753e973a6fb38493a3R30 works w/o special handling.

I could capture the `boolean` and have a single `table(rs)` at the very end, but I really wish I didn't have to.

Short of wrapping a anonymous procedure around the real procedure, I couldn't come up with a way to "rename" the output column from the proc (the column name of the result is always the name of the procedure itself).